build/model_dist.py

```diff
@@ -0,0 +1,277 @@
+# LICENSE file in the root directory of this source tree.
+from pathlib import Path
+from typing import Dict, Optional
+
+import torch
+import torch.nn as nn
+
+from torch import Tensor
+from torch.nn import functional as F
+from torch.distributed._tensor import DTensor, Replicate
+from torch.distributed.device_mesh import _mesh_resources, DeviceMesh
+from torch.distributed.tensor.parallel import parallelize_module, ColwiseParallel, RowwiseParallel
+
+from build.utils import find_multiple
+
+from build.model import TransformerArgs, KVCache, apply_rotary_emb, precompute_freqs_cis
+
+config_path = Path(f"{str(Path(__file__).parent)}/known_model_params")
+
+
+# Use DTensor as output, by default
+Colwise = ColwiseParallel(use_local_output=False)
+Rowwise = RowwiseParallel(use_local_output=False)
+
+# Device mesh context
+device_mesh = None
+
+
+class Transformer(nn.Module):
+    def __init__(self, config: TransformerArgs) -> None:
+        super().__init__()
+        self.config = config
+
+        # Get device mesh
+        global device_mesh
+        if device_mesh is None:
+            device_mesh = _mesh_resources.get_current_mesh()
+
+        tok_embeddings = nn.Embedding(config.vocab_size, config.dim)
+        self.tok_embeddings = parallelize_module(
+            tok_embeddings,
+            device_mesh,
+            RowwiseParallel(input_layouts=Replicate()),
+        )
+        self.layers = nn.ModuleList(
+            TransformerBlock(config) for _ in range(config.n_layers)
+        )
+        self.norm = RMSNorm(config.dim, eps=config.norm_eps)
+        self.output = nn.Linear(config.dim, config.vocab_size, bias=False)
+
+        # self.freqs_cis: Optional[Tensor] = None
+        # self.mask_cache: Optional[Tensor] = None
+        self.max_batch_size = -1
+        self.max_seq_length = -1
+
+    def setup_caches(self, max_batch_size, max_seq_length):
+        if (
+            self.max_seq_length >= max_seq_length
+            and self.max_batch_size >= max_batch_size
+        ):
+            return
+        head_dim = self.config.dim // self.config.n_heads
+        max_seq_length = find_multiple(max_seq_length, 8)
+        self.max_seq_length = max_seq_length
+        self.max_batch_size = max_batch_size
+        for b in self.layers:
+            b.attention.kv_cache = KVCache(
+                max_batch_size, max_seq_length, self.config.n_local_heads, head_dim
+            )
+
+        freqs_cis = precompute_freqs_cis(
+            self.config.dim // self.config.n_heads,
+            self.config.block_size * 2,
+            self.config.rope_base,
+            use_scaled = self.config.use_scaled_rope,
+        )
+        self.register_buffer("freqs_cis", freqs_cis, persistent=True)
+        causal_mask = torch.tril(
+            torch.ones(self.max_seq_length, self.max_seq_length, dtype=torch.bool)
+        )
+        self.register_buffer("causal_mask", causal_mask, persistent=True)
```

Comment:
same q as above re: should persistent =True?

Reply:
coming from original model too